File: order-service/src/test/java/com/catius/order/service/OrderServiceTest.java
package com.catius.order.service;

import com.catius.order.controller.dto.request.OrderItemRequest;
import com.catius.order.controller.dto.request.OrderRequest;
import com.catius.order.controller.dto.response.OrderResponse;
import com.catius.order.domain.Order;
import com.catius.order.domain.OrderItem;
import com.catius.order.domain.OrderStatus;
import com.catius.order.event.OrderSagaOrchestrator;
import com.catius.order.exception.OrderNotFoundException;
import com.catius.order.repository.OrderRepository;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.mockito.InjectMocks;
import org.mockito.Mock;
import org.mockito.junit.jupiter.MockitoExtension;
import org.springframework.test.util.ReflectionTestUtils;

import java.util.List;
import java.util.Optional;

import static org.assertj.core.api.Assertions.*;
import static org.mockito.ArgumentMatchers.any;
import static org.mockito.BDDMockito.*;

@ExtendWith(MockitoExtension.class)
class OrderServiceTest {

    @Mock
    private OrderRepository orderRepository;

    @Mock
    private OrderSagaOrchestrator orderSagaOrchestrator;

    @InjectMocks
    private OrderService orderService;

    @Test
    @DisplayName("주문 생성 - 성공")
    void createOrder_성공() {
        given(orderRepository.save(any(Order.class))).willAnswer(inv -> {
            Order saved = inv.getArgument(0);
            ReflectionTestUtils.setField(saved, "id", 1L);
            return saved;
        });
        willDoNothing().given(orderSagaOrchestrator).execute(any(Order.class));

        OrderRequest request = new OrderRequest(1L, List.of(new OrderItemRequest(1001L, 2)));
        OrderResponse response = orderService.createOrder(request);

        assertThat(response.id()).isEqualTo(1L);
        assertThat(response.items()).hasSize(1);
        assertThat(response.items().get(0).productId()).isEqualTo(1001L);
        assertThat(response.items().get(0).quantity()).isEqualTo(2);
        assertThat(response.status()).isEqualTo(OrderStatus.PENDING);
        then(orderSagaOrchestrator).should().execute(any(Order.class));
    }

    @Test
    @DisplayName("주문 조회 - 성공")
    void findById_성공() {
        Order order = Order.create(1L, List.of(OrderItem.of(1001L, 2)));
        given(orderRepository.findById(1L)).willReturn(Optional.of(order));

        OrderResponse response = orderService.findById(1L);

        assertThat(response.items().get(0).productId()).isEqualTo(1001L);
        assertThat(response.status()).isEqualTo(OrderStatus.PENDING);
    }

    @Test
    @DisplayName("주문 조회 - 없는 주문 → 404")
    void findById_없는주문_404() {
        given(orderRepository.findById(999L)).willReturn(Optional.empty());

        assertThatThrownBy(() -> orderService.findById(999L))
                .isInstanceOf(OrderNotFoundException.class)
                .hasMessage("Order not found: 999");
    }
}

File: order-service/src/test/kotlin/com/catius/order/service/OrderServiceTest.kt
package com.catius.order.service

import com.catius.order.domain.Order
import com.catius.order.domain.OrderDomainEvent
import com.catius.order.domain.OrderItem
import com.catius.order.domain.OrderStatus
import com.catius.order.domain.port.InventoryClient
import com.catius.order.domain.port.OrderEventPublisher
import com.catius.order.domain.port.OrderRepository
import com.catius.order.domain.port.ReserveOutcome
import com.catius.order.domain.usecase.command.OrderCommand
import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.DisplayName
import org.junit.jupiter.api.Nested
import org.junit.jupiter.api.Test
import org.mockito.ArgumentMatchers.any
import org.mockito.ArgumentMatchers.eq
import org.mockito.BDDMockito.given
import org.mockito.Mockito.mock
import org.mockito.Mockito.never
import org.mockito.Mockito.times
import org.mockito.Mockito.verify
import org.springframework.transaction.TransactionStatus
import org.springframework.transaction.support.TransactionCallback
import org.springframework.transaction.support.TransactionTemplate
import java.time.Clock
import java.time.Instant
import java.time.ZoneOffset

class OrderServiceTest {

    private val now = Instant.parse("2026-04-28T00:00:00Z")
    private val fixedClock = Clock.fixed(now, ZoneOffset.UTC)

    private lateinit var orderRepository: OrderRepository
    private lateinit var inventoryClient: InventoryClient
    private lateinit var orderEventPublisher: OrderEventPublisher
    private lateinit var transactionTemplate: TransactionTemplate
    private lateinit var service: OrderService

    private fun <T> anyK(): T {
        any<T>()
        @Suppress("UNCHECKED_CAST")
        return null as T
    }

    @BeforeEach
    fun setup() {
        orderRepository = mock(OrderRepository::class.java)
        inventoryClient = mock(InventoryClient::class.java)
        orderEventPublisher = mock(OrderEventPublisher::class.java)
        transactionTemplate = mock(TransactionTemplate::class.java)
        service = OrderService(orderRepository, inventoryClient, orderEventPublisher, fixedClock, transactionTemplate)

        // Mock transactionTemplate.execute to call the callback immediately
        given(transactionTemplate.execute(anyK<TransactionCallback<Any>>())).willAnswer {
            val callback = it.getArgument<TransactionCallback<Any>>(0)
            callback.doInTransaction(mock(TransactionStatus::class.java))
        }
    }

    @Nested
    @DisplayName("주문 생성 (CreateOrder)")
    inner class CreateOrder {

        @Test
        @DisplayName("재고 예약 성공 시 주문이 CONFIRMED 가 되고 이벤트를 발행한다")
        fun create_success() {
            // given
            val command = OrderCommand.CreateOrderCommand(
                customerId = 1L,
                items = listOf(OrderCommand.CreateOrderCommand.Item(productId = 1001L, quantity = 2)),
            )
            val pendingOrder = Order.create(1L, listOf(OrderItem(1001L, 2)), now).copy(id = 42L)
            val confirmedOrder = pendingOrder.confirm(now)

            given(orderRepository.save(anyK())).willReturn(pendingOrder, confirmedOrder)
            given(orderRepository.findById(42L)).willReturn(pendingOrder)
            given(inventoryClient.reserve(eq(42L), anyK())).willReturn(ReserveOutcome.Success)

            // when
            val result = service.create(command)

            // then
            assertThat(result.id).isEqualTo(42L)
            assertThat(result.status).isEqualTo(OrderStatus.CONFIRMED)

            verify(orderRepository, times(2)).save(anyK())
            verify(orderRepository).findById(42L)
            verify(inventoryClient).reserve(eq(42L), anyK())
            verify(orderEventPublisher).publish(anyK())
        }

        @Test
        @DisplayName("재고 부족 시 주문이 FAILED 가 되고 release 는 호출하지 않는다")
        fun create_insufficient_stock() {
            // given
            val command = OrderCommand.CreateOrderCommand(
                customerId = 1L,
                items = listOf(OrderCommand.CreateOrderCommand.Item(productId = 1001L, quantity = 2)),
            )
            val pendingOrder = Order.create(1L, listOf(OrderItem(1001L, 2)), now).copy(id = 42L)
            val failedOrder = pendingOrder.fail(now)

            given(orderRepository.save(anyK())).willReturn(pendingOrder, failedOrder)
            given(orderRepository.findById(42L)).willReturn(pendingOrder)
            given(inventoryClient.reserve(eq(42L), anyK())).willReturn(ReserveOutcome.InsufficientStock(1001L))

            // when
            val result = service.create(command)

            // then
            assertThat(result.status).isEqualTo(OrderStatus.FAILED)
            verify(inventoryClient, never()).release(any(Long::class.java), anyK())
            verify(orderEventPublisher, never()).publish(anyK())
        }

        @Test
        @DisplayName("재고 서버 장애(Unavailable) 시 release 를 호출하고 주문이 FAILED 가 된다")
        fun create_unavailable() {
            // given
            val command = OrderCommand.CreateOrderCommand(
                customerId = 1L,
                items = listOf(OrderCommand.CreateOrderCommand.Item(productId = 1001L, quantity = 2)),
            )
            val pendingOrder = Order.create(1L, listOf(OrderItem(1001L, 2)), now).copy(id = 42L)
            val failedOrder = pendingOrder.fail(now)

            given(orderRepository.save(anyK())).willReturn(pendingOrder, failedOrder)
            given(orderRepository.findById(42L)).willReturn(pendingOrder)
            given(inventoryClient.reserve(eq(42L), anyK())).willReturn(ReserveOutcome.Unavailable)

            // when
            val result = service.create(command)

            // then
            assertThat(result.status).isEqualTo(OrderStatus.FAILED)
            verify(inventoryClient).release(eq(42L), anyK())
            verify(orderEventPublisher, never()).publish(anyK())
        }
    }
}
